File: OneCodeTeam/Claim-aware-WCF-WebServicevs10-17865/[VB]-Claim-aware WCF WebService[vs10] (VBSharePointCallClaimsAwareWCF)/VB/VBSharePointCallClaimsAwareWCF/Client/CustomServiceApplicationProxy.vb
'************************** Module Header ******************************\
' Module Name:    CustomServiceApplicationProxy.vb
' Project:        VBSharePointCallClaimsAwareWCF
' Copyright (c) Microsoft Corporation
'
' This class is used to define ServiceApplicationProxy
' 
' This source is subject to the Microsoft Public License.
' See http://www.microsoft.com/en-us/openness/licenses.aspx#MPL.
' All other rights reserved.
' 
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
' WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
'\****************************************************************************


Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports Microsoft.SharePoint.Administration
Imports System.ServiceModel
Imports Microsoft.SharePoint
Imports Microsoft.SharePoint.Utilities
Imports System.ServiceModel.Configuration
Imports System.ServiceModel.Channels
Imports System.Configuration
Imports System.Runtime.InteropServices

Namespace CustomService
    <IisWebServiceApplicationProxyBackupBehavior()> _
    <Guid("D409CE5C-19BD-4269-8B16-8F3EDCCB1039")> _
    Public NotInheritable Class CustomServiceApplicationProxy
        Inherits SPIisWebServiceApplicationProxy
        Private _channelFactory As ChannelFactory(Of ICustomServiceContract)
        Private _channelFactoryLock As New Object()
        Private _endpointConfigName As String

        <Persisted()> _
        Private _loadBalancer As SPServiceLoadBalancer

#Region "constructors"
        Public Sub New()
        End Sub

        Public Sub New(ByVal name As String, ByVal serviceProxy As CustomServiceProxy, ByVal serviceAppAddress As Uri)
            MyBase.New(name, serviceProxy, serviceAppAddress)
            ' create a new round robin load balancer
            _loadBalancer = New SPRoundRobinServiceLoadBalancer(serviceAppAddress)
        End Sub
#End Region

#Region "app proxy infrastructure"
        Private Function CreateChannelFactory(Of T)(ByVal endpointConfigName As String) As ChannelFactory(Of T)
            ' open client.config
            Dim clientConfigPath As String = SPUtility.GetGenericSetupPath("WebClients\CustomService")
            Dim clientConfig As Configuration = OpenClientConfiguration(clientConfigPath)
            Dim factory As New ConfigurationChannelFactory(Of T)(endpointConfigName, clientConfig, Nothing)

            ' configure the channel factory for IDFx claims auth
            factory.ConfigureCredentials(SPServiceAuthenticationMode.Claims)

            Return factory
        End Function

        Friend Delegate Sub CodeToRunOnApplicationProxy(ByVal appProxy As CustomServiceApplicationProxy)
        Friend Shared Sub Invoke(ByVal serviceContext As SPServiceContext, ByVal codeBlock As CodeToRunOnApplicationProxy)
            If serviceContext Is Nothing Then
                Throw New ArgumentNullException("serviceContext")
            End If

            ' get sample service app proxy from context
            Dim proxy As CustomServiceApplicationProxy = DirectCast(serviceContext.GetDefaultProxy(GetType(CustomServiceApplicationProxy)), CustomServiceApplicationProxy)
            If proxy Is Nothing Then
                Throw New InvalidOperationException("Custom service application proxy not found.")
            End If

            ' ensure the current service context is correctly set
            Using New SPServiceContextScope(serviceContext)
                ' execute the delegate
                codeBlock(proxy)
            End Using
        End Sub

        Private Delegate Sub CodeToRunOnChannel(ByVal serviceContract As ICustomServiceContract)
        Private Sub ExecuteOnChannel(ByVal operationName As String, ByVal options As ExecuteOptions, ByVal codeBlock As CodeToRunOnChannel)
            Using New SPMonitoredScope("ExecuteOnChannel:" & operationName)
                Dim loadBalancerCtx As SPServiceLoadBalancerContext = _loadBalancer.BeginOperation()
                Try
                    ' get a channel to the service app endpoint
                    Dim channel As IChannel = DirectCast(GetChannel(loadBalancerCtx.EndpointAddress, options), IChannel)
                    Try
                        ' execute the delegate
                        codeBlock(DirectCast(channel, ICustomServiceContract))

                        'close the channel
                        channel.Close()
                    Catch generatedExceptionName As TimeoutException
                        loadBalancerCtx.Status = SPServiceLoadBalancerStatus.Failed
                        Throw
                    Catch generatedExceptionName As EndpointNotFoundException
                        loadBalancerCtx.Status = SPServiceLoadBalancerStatus.Failed
                        Throw
                    Finally
                        If channel.State <> CommunicationState.Closed Then
                            channel.Abort()
                        End If
                    End Try
                Finally
                    loadBalancerCtx.EndOperation()
                End Try
            End Using
        End Sub

        ''' <summary>
        ''' Gets the endpoint configuration name for a given endpoint address.
        ''' </summary>
        ''' <param name="address">Endpoint address.</param>
        ''' <returns>The endpoint configuration name.</returns>
        ''' <remarks>The returned endpont must match one of the endpoint element names in the client.config file.</remarks>
        Private Function GetEndpointConfigurationName(ByVal address As Uri) As String
            If address Is Nothing Then
                Throw New ArgumentNullException("address")
            End If

            Dim configName As String

            If address.Scheme = Uri.UriSchemeNetTcp Then
                If address.AbsolutePath.EndsWith("/secure", StringComparison.OrdinalIgnoreCase) Then
                    configName = "tcp-ssl"
                Else
                    configName = "tcp"
                End If
            ElseIf address.Scheme = Uri.UriSchemeHttps Then
                configName = "https"
            ElseIf address.Scheme = Uri.UriSchemeHttp Then
                configName = "http"
            Else
                Throw New NotSupportedException("Unsupported endpoint address.")
            End If

            Return configName
        End Function

        Private Function GetChannel(ByVal address As Uri, ByVal options As ExecuteOptions) As ICustomServiceContract
            ' get the endpoint config name
            Dim endpointConfigName As String = GetEndpointConfigurationName(address)

            ' check for a cached channel factory for the endpoint config
            If (_channelFactory Is Nothing) OrElse (endpointConfigName <> _endpointConfigName) Then
                SyncLock _channelFactoryLock
                    ' create a channel factory using the default endpoing config name
                    _channelFactory = CreateChannelFactory(Of ICustomServiceContract)(endpointConfigName)

                    ' cache the endpoint config name
                    _endpointConfigName = endpointConfigName
                End SyncLock
            End If

            Dim channel As ICustomServiceContract

            If ExecuteOptions.AsProcess = (options And ExecuteOptions.AsProcess) Then
                ' create a channel that acts as the service's process user when authenticating with the service
                channel = _channelFactory.CreateChannelAsProcess(New EndpointAddress(address))
            Else
                ' create a channel that acts as the loged on user when authenticating with the service
                channel = _channelFactory.CreateChannelActingAsLoggedOnUser(New EndpointAddress(address))
            End If

            ' create a channel from the factory
            Return channel
        End Function
#End Region

        Public Overrides ReadOnly Property TypeName() As String
            Get
                Return "Custom Service Application Proxy"
            End Get
        End Property

        Public Overrides Sub Provision()
            _loadBalancer.Provision()
            MyBase.Provision()
            Me.Update()
        End Sub

        Public Overrides Sub Unprovision(ByVal deleteData As Boolean)
            _loadBalancer.Unprovision()
            MyBase.Unprovision(deleteData)
            Me.Update()
        End Sub

        ''' <summary>
        ''' Client method executed on WFE (front-end), for example, by a web part.
        ''' </summary>
#Region "service app methods"
        Public Function Add(ByVal a As Integer, ByVal b As Integer, ByVal options As ExecuteOptions) As Integer
            Dim result As Integer = 0

            ExecuteOnChannel("Add", options, Function(proxy) InlineAssignHelper(result, proxy.Add(a, b)))

            Return result
        End Function

        Public Function HelloWorld(ByVal options As ExecuteOptions) As String
            Dim result As String = String.Empty

            ' Execute the service application method
            Me.ExecuteOnChannel("HelloWorld", options, Function(proxy) InlineAssignHelper(result, proxy.HelloWorld()))

            Return result
        End Function

        Private Shared Function InlineAssignHelper(Of T)(ByRef target As T, ByVal value As T) As T
            target = value
            Return value
        End Function
#End Region
    End Class
End Namespace

File: OneCodeTeam/Claim-aware-WCF-WebServicevs10-17866/[C#]-Claim-aware WCF WebService[vs10] (CSSharePointCallClaimsAwareWCF)/C#/CSSharePointCallClaimsAwareWCF/Client/CustomServiceApplicationProxy.cs
/****************************** Module Header ******************************\
* Module Name:    CustomServiceApplicationProxy.cs
* Project:        CSSharePointCallClaimsAwareWCF
* Copyright (c) Microsoft Corporation
*
* This class is used to define ServiceApplicationProxy
* 
* This source is subject to the Microsoft Public License.
* See http://www.microsoft.com/en-us/openness/licenses.aspx#MPL.
* All other rights reserved.
* 
* THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
* EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
* WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
\*****************************************************************************/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.SharePoint.Administration;
using System.ServiceModel;
using Microsoft.SharePoint;
using Microsoft.SharePoint.Utilities;
using System.ServiceModel.Configuration;
using System.ServiceModel.Channels;
using System.Configuration;
using System.Runtime.InteropServices;

namespace CustomService
{
    [IisWebServiceApplicationProxyBackupBehavior]
    [Guid("D409CE5C-19BD-4269-8B16-8F3EDCCB1039")]
    public sealed class CustomServiceApplicationProxy : SPIisWebServiceApplicationProxy
    {
        private ChannelFactory<ICustomServiceContract> _channelFactory;
        private object _channelFactoryLock = new object();
        private string _endpointConfigName;

        [Persisted]
        private SPServiceLoadBalancer _loadBalancer;

        #region constructors
        public CustomServiceApplicationProxy() { }

        public CustomServiceApplicationProxy(string name, CustomServiceProxy serviceProxy, Uri serviceAppAddress)
            : base(name, serviceProxy, serviceAppAddress)
        {
            // create a new round robin load balancer
            _loadBalancer = new SPRoundRobinServiceLoadBalancer(serviceAppAddress);
        }
        #endregion

        #region app proxy infrastructure
        private ChannelFactory<T> CreateChannelFactory<T>(string endpointConfigName)
        {
            // open client.config
            string clientConfigPath = SPUtility.GetGenericSetupPath(@"WebClients\CustomService");
            Configuration clientConfig = OpenClientConfiguration(clientConfigPath);
            ConfigurationChannelFactory<T> factory = new ConfigurationChannelFactory<T>(endpointConfigName, clientConfig, null);

            // configure the channel factory for IDFx claims auth
            factory.ConfigureCredentials(SPServiceAuthenticationMode.Claims);

            return factory;
        }

        internal delegate void CodeToRunOnApplicationProxy(CustomServiceApplicationProxy appProxy);
        internal static void Invoke(SPServiceContext serviceContext, CodeToRunOnApplicationProxy codeBlock)
        {
            if (serviceContext == null)
                throw new ArgumentNullException("serviceContext");

            // get sample service app proxy from context
            CustomServiceApplicationProxy proxy = (CustomServiceApplicationProxy)serviceContext.GetDefaultProxy(typeof(CustomServiceApplicationProxy));
            if (proxy == null)
                throw new InvalidOperationException("Custom service application proxy not found.");

            // ensure the current service context is correctly set
            using (new SPServiceContextScope(serviceContext))
            {
                // execute the delegate
                codeBlock(proxy);
            }
        }

        private delegate void CodeToRunOnChannel(ICustomServiceContract serviceContract);
        private void ExecuteOnChannel(string operationName, ExecuteOptions options, CodeToRunOnChannel codeBlock)
        {
            using (new SPMonitoredScope("ExecuteOnChannel:" + operationName))
            {
                SPServiceLoadBalancerContext loadBalancerCtx = _loadBalancer.BeginOperation();
                try
                {
                    // get a channel to the service app endpoint
                    IChannel channel = (IChannel)GetChannel(loadBalancerCtx.EndpointAddress, options);
                    try
                    {
                        // execute the delegate
                        codeBlock((ICustomServiceContract)channel);

                        //close the channel
                        channel.Close();
                    }
                    catch (TimeoutException)
                    {
                        loadBalancerCtx.Status = SPServiceLoadBalancerStatus.Failed;
                        throw;
                    }
                    catch (EndpointNotFoundException)
                    {
                        loadBalancerCtx.Status = SPServiceLoadBalancerStatus.Failed;
                        throw;
                    }
                    finally
                    {
                        if (channel.State != CommunicationState.Closed)
                            channel.Abort();
                    }
                }
                finally
                {
                    loadBalancerCtx.EndOperation();
                }
            }
        }

        /// <summary>
        /// Gets the endpoint configuration name for a given endpoint address.
        /// </summary>
        /// <param name="address">Endpoint address.</param>
        /// <returns>The endpoint configuration name.</returns>
        /// <remarks>The returned endpont must match one of the endpoint element names in the client.config file.</remarks>
        private string GetEndpointConfigurationName(Uri address)
        {
            if (address == null)
                throw new ArgumentNullException("address");

            string configName;

            if (address.Scheme == Uri.UriSchemeNetTcp)
            {
                if (address.AbsolutePath.EndsWith("/secure", StringComparison.OrdinalIgnoreCase))
                    configName = "tcp-ssl";
                else configName = "tcp";
            }
            else if (address.Scheme == Uri.UriSchemeHttps)
                configName = "https";
            else if (address.Scheme == Uri.UriSchemeHttp)
                configName = "http";
            else
                throw new NotSupportedException("Unsupported endpoint address.");

            return configName;
        }

        private ICustomServiceContract GetChannel(Uri address, ExecuteOptions options)
        {
            // get the endpoint config name
            string endpointConfigName = GetEndpointConfigurationName(address);

            // check for a cached channel factory for the endpoint config
            if ((_channelFactory == null) || (endpointConfigName != _endpointConfigName))
            {
                lock (_channelFactoryLock)
                {
                    // create a channel factory using the default endpoing config name
                    _channelFactory = CreateChannelFactory<ICustomServiceContract>(endpointConfigName);

                    // cache the endpoint config name
                    _endpointConfigName = endpointConfigName;
                }
            }

            ICustomServiceContract channel;

            if (ExecuteOptions.AsProcess == (options & ExecuteOptions.AsProcess))
                // create a channel that acts as the service's process user when authenticating with the service
                channel = _channelFactory.CreateChannelAsProcess<ICustomServiceContract>(new EndpointAddress(address));
            else
                // create a channel that acts as the loged on user when authenticating with the service
                channel = _channelFactory.CreateChannelActingAsLoggedOnUser<ICustomServiceContract>(new EndpointAddress(address));

            // create a channel from the factory
            return channel;
        }
        #endregion

        public override string TypeName
        {
            get { return "Custom Service Application Proxy"; }
        }

        public override void Provision()
        {
            _loadBalancer.Provision();
            base.Provision();
            this.Update();
        }

        public override void Unprovision(bool deleteData)
        {
            _loadBalancer.Unprovision();
            base.Unprovision(deleteData);
            this.Update();
        }

        /// <summary>
        /// Client method executed on WFE (front-end), for example, by a web part.
        /// </summary>
        #region service app methods
        public int Add(int a, int b, ExecuteOptions options)
        {
            int result = 0;

            ExecuteOnChannel("Add", options, delegate(ICustomServiceContract channel)
            {
                result = channel.Add(a, b);
            });

            return result;
        }

        public string HelloWorld(ExecuteOptions options)
        {
            string result = string.Empty;

            // Execute the service application method
            this.ExecuteOnChannel("HelloWorld", options, delegate(ICustomServiceContract channel)
            {
                result = channel.HelloWorld();
            });               

            return result;
        }
        #endregion
    }
}